tokio/src/runtime/thread_pool/queue.rs

```diff
@@ -0,0 +1,550 @@
+            };
+
+            self.inner.buffer[i_idx].with_mut(|ptr| unsafe {
+                let ptr = (*ptr).as_ptr();
+                *(*ptr).header().queue_next.get() = Some(next);
+            });
+        }
+
+        let head = self.inner.buffer[head as usize & MASK]
+            .with(|ptr| unsafe { ptr::read((*ptr).as_ptr()) });
+
+        // Push the tasks onto the inject queue
+        inject.push_batch(head, task, BATCH_LEN);
+
+        Ok(())
+    }
+
+    /// Pops a task from the local queue.
+    pub(super) fn pop(&mut self) -> Option<task::Notified<T>> {
+        // If a task is available in the FIFO slot, return that.
+        if let Some(task) = self.next.take() {
+            return Some(task);
+        }
+
+        loop {
+            let head = self.inner.head.load(Acquire);
+
+            // safety: this is the **only** thread that updates this cell.
+            let tail = unsafe { self.inner.tail.unsync_load() };
+
+            if head == tail {
+                // queue is empty
+                return None;
+            }
+
+            // Map the head position to a slot index.
+            let idx = head as usize & MASK;
+
+            let task = self.inner.buffer[idx].with(|ptr| {
+                // Tentatively read the task at the head position. Note that we
+                // have not yet claimed the task.
+                //
+                unsafe { ptr::read(ptr) }
+            });
+
+            // Attempt to claim the task read above.
+            let actual = self
+                .inner
+                .head
+                .compare_and_swap(head, head.wrapping_add(1), Release);
+
+            if actual == head {
+                return Some(unsafe { task.assume_init() });
+            }
+
+            atomic::spin_loop_hint();
+        }
+    }
+}
+
+impl<T> Steal<T> {
+    pub(super) fn is_empty(&self) -> bool {
+        self.0.is_empty()
+    }
+
+    /// Steals half the tasks from self and place them into `dst`.
+    pub(super) fn steal_into(&self, dst: &mut Local<T>) -> Option<task::Notified<T>> {
+        let dst_tail = unsafe { dst.inner.tail.unsync_load() };
+
+        // Steal the tasks into `dst`'s buffer. This does not yet expose the
+        // tasks in `dst`.
+        let mut n = self.steal_into2(dst, dst_tail);
+
+        if n == 0 {
+            // No tasks were stolen
+            return None;
+        }
+
+        // We are returning a task here
+        n -= 1;
+
+        let ret_pos = dst_tail.wrapping_add(n);
+        let ret_idx = ret_pos as usize & MASK;
+
+        let ret = dst.inner.buffer[ret_idx].with(|ptr| unsafe { ptr::read((*ptr).as_ptr()) });
+
+        if n == 0 {
+            // The `dst` queue is empty, but a single task was stolen
+            return Some(ret);
+        }
+
+        // Synchronize with stealers
+        let dst_head = dst.inner.head.load(Acquire);
+
+        assert!(dst_tail.wrapping_sub(dst_head) + n <= LOCAL_QUEUE_CAPACITY as u32);
```

Comment:
Also a debug assert? This one is also subtle enough that I think a brief comment would be good.